### ISSUE TITLE
Calculated types or custom type hint and type checkers

### ISSUE THREAD

**Issue by chatgpt-api-cn** [erg-lang/erg#445]
There's been a long history of attempts on typechecking libraries like `pytorch` and `numpy`. One major issue is that passing different parameters to these constructors create different types (tensors with traits (matrix multiplication) and tensor functions). There's some relationship between definition of a convolutional network and what type of value (usually not a fixed type, but rather a set of types (infinite)) it can accept and return.

So I propose or want to know how we can let Erg to solve this long-standing problem, by letting the static typechecker to know the type of `torch.nn.Conv2d` and emit errors before execution? Help can be found by:

```python
import torch
help(torch.nn.Conv2d)
```

**Comment by mtshiba** [comment]
Yes, in fact one of Erg's motivations for introducing dependent types is to be able to verify array and tensor transformations at compile time.
I am currently trying to have the information of `numpy.ndarray.shape` in the type as a start. Stay tuned!
